File: src/Python/find_matrix.py
###########################################################
# @Author: HoodUSSEnterprise
# @Date: 2026-06-22 13:08:20
# @LastEditors: HoodUSSEnterprise
# @LastEditTime: 2026-06-22 13:09:16
# @FilePath: \asm_matrix_benchmark\src\Python\find_matrix.py
# @Description: find matrix python code
###########################################################
from typing import Any, Union

from .base_matrix import Matrix


###########################################################
# @description: find elem in matrix
# @param {Any} m
# @param {Union} v
# @param {*} float
# @return {*}
###########################################################
def find_elem(m: Any, v: Union[int, float]) -> tuple[int, int]:
    if m is None or not isinstance(m, Matrix):
        print("Invalid param")
        return -1, -1

    for i in range(m.rows):
        for j in range(m.cols):
            if m.data[i * m.cols + j] == v:
                return i, j

    print(f"No find elem {v}")
    return -1, -1

File: src/C++/float/find_matrix.cpp
/************************************************************
@Author: HoodUSSEnterprise
@Date: 2026-06-23 17:36:46
@LastEditors: HoodUSSEnterprise
@LastEditTime: 2026-06-26 09:04:32
@FilePath: \asm_matrix_benchmark\src\C++\float\find_matrix.cpp
@Description: find matrix c++ code
*************************************************************/

#include "matrix_float_cpp.h"

/***********************************************************
@description:Searches for the first occurrence of a specified value in the matrix using row-major order
@param {float} value
@return {*}
*************************************************************/
std::pair<int, int> MatrixFloat::find_elem(float value) const
{
    // Scan in row-major order for the first occurrence
    for (size_t i = 0; i < rows; i++)
    {
        for (size_t j = 0; j < cols; j++)
        {
            if (data[i * cols + j] == value)
            {
                return std::pair<int, int>(static_cast<int>(i), static_cast<int>(j));
            }
        }
    }
    // Return (-1, -1) if not found
    return std::pair<int, int>(-1, -1);
}
